### PR TITLE
Make TaprooBuilder::finalize able to return keyspend only

### PR DESCRIPTION
**Pull request by JeremyRubin**

### CODE REVIEW

File: src/util/taproot.rs [hunk 1]
```diff
@@ -467,20 +467,21 @@ impl TaprootBuilder {
...
         internal_key: UntweakedPublicKey,
     ) -> Result<TaprootSpendInfo, TaprootBuilderError> {
-        if self.branch.len() > 1 {
-            return Err(TaprootBuilderError::IncompleteTree);
+        match self.branch.pop() {
+            None => Ok(TaprootSpendInfo::new_key_spend(secp, internal_key, None)),
+            Some(Some(node)) => {
+                Ok(TaprootSpendInfo::from_node_info(secp, internal_key, node))
+            }
+            _ => Err(TaprootBuilderError::IncompleteTree),
```
Comment (sanket1729) [inline src/util/taproot.rs:482]: I don't like this because this modifies the Builder in the error case.

**Review by sanket1729 — COMMENTED** [review]
One small comment

File: src/util/taproot.rs (hunk 1 above)
Comment (apoelstra) [reply src/util/taproot.rs:482]: Agreed -- can you change this to `push` the data back onto `self.branch` in this case?

**Comment by JeremyRubin** [comment]
Why is this important if the function takes ownership of the value? It
would be dropped in the error, no?

Happy to fixup soon if it matters, mostly asking to see if I'm missing
something else.

**Comment by tcharding** [comment]
Ha! I missed the ownership thing during review as well, perhaps a code comment in the function mentioning it?

**Comment by sanket1729** [comment]
> Why is this important if the function takes ownership of the value?

The inner API `from_node_info` needs to consume the value. I think it is idiomatic of the builder pattern to consume values and generate new values.

>  It would be dropped in the error, no?

Yes, but this was in the previous case when the builder was empty. Perhaps, the correct thing to do here(and other builder APIs) is to return to the previous builder state in the error value.

**Comment by JeremyRubin** [comment]
@sanket1729 that sounds like a bigger change because then it's a breaking API change vs a bug fix that can be minor released.

Can we merge this patch for a point release as-is and we can open a new issue for the above (change to a continuation that lets you continue?)

**Review by sanket1729 — APPROVED** [review]
ACK 7969b7a43e0f1415f60baa6729a3f4b5963718ed

**Review by apoelstra — APPROVED** [review]
ACK 7969b7a43e0f1415f60baa6729a3f4b5963718ed

I missed the ownership thing as well :).

TBH I don't really follow the discussion here. Will hold off on merging until @sanket1729 agrees that it's resolved. But I'm fine to pull this into the release. It is technically adding extra functionality, which we aren't supposed to do, but it's small.

**Comment by JeremyRubin** [comment]
I think the resolution is that we'll do a breaking API change later to make
finalize return an error + state to keep working on it if finalize failed,
but we'll do that after a non breaking change to just fix the 'non
totality' of the current builder, which should go into the release.

By non totality, I mean there are reasonable taprootspendinfos that cannot
be built

**Comment by sanket1729** [comment]
> I mean there are reasonable taprootspendinfos that cannot be built

Can you point to an example?

**Comment by JeremyRubin** [comment]
Probably I should have included a test showing this.

**Comment by JeremyRubin** [comment]
Yes the example being a key spend only taprootspendinfo cannot be built
from taprootbuilder before this patch, after the patch it can be. I.e.,
should all taprootspendinfos be reachable by the builder?

**Pull request merged**